File: deepspace/Dice.rb
#encondig:utf-8

# WIP -- Documentation needed
# Class to take random decisions
class Dice 
	def initialize()
		@NHANGARSPROB = 0.25
		@NSHIELDSPROB = 0.25
		@NWEAPONSPROB = 0.33
		@FIRSTSHOTPROB = 0.5

		# Random number generator
		@generator = Random.new()

	end

	# Description:
	# 	Determines the number of hangars that a space station will recieve
	#	upon creation. This value is calculated based on a random value between
	#	0 and 1, and the return value depends on NHANGARSPROB's probability:
	#		- 0 if random_value <= NHANGARSPROB
	#		- 1 otherwise
	# Return
	#	Integer, 0 or 1
	def initWithNHangars
		if @generator.rand <= @NHANGARSPROB
			return 0
		else
			return 1
		end

		# Security check
		puts "WARNING! UNEXPECTED CONDITION AT Dice.initWithNHangars"
	end

	
	# Description:
	#	Determines the number of weapons that a space station will recieve
	#	upon creation. This value is calculated based on a random value between
	#	0 and 1, and the return value depends on NWEAPONSPROB's probability:
	#		- 1 if random_value <= NWEAPONSPROB
	#		- 2 else if random_value <= 2*NWEAPONSPROB
	#		- 3 otherwise
	# Return
	#	Integer, 0 or 1
	#int initWithNWeapons: devuelve 1 con una probabilidad de NWEAPONSPROB, 2 con la misma
	#probabilidad y 3 con una probabilidad de (1-2* NWEAPONSPROB). Este método determina el
	#número de armas que recibirá una estación espacial al ser creada.
	def initWithNWeapons
		randval = @generator.rand
		if randval <= @NWEAPONSPROB
			return 1
		elsif randval > @NWEAPONSPROB and randval <= @NWEAPONSPROB*2
			return 2
		else
			return 3
		end

		# Security check
		puts "WARNING! UNEXPECTED CONDITION AT Dice.initWIthNWeapons"
	end

	# Description:
	#	Determines amount of shield enhancers that a space station will recieve
	#	upon creation. This value is calculated based on a random value between
	#	0 and 1, and the return value depends on NSHIELDSPROB's probability:
	#		- 0 if random_value <= NSHIELDSPROB
	#		- 1 otherwise
	# Return
	#	Integer, 0 or 1
	def initWithNShields
		if @generator.rand <= @NSHIELDSPROB
			return 0
		else
			return 1
		end

		# Security check
		puts "WARNING! UNEXPECTED CONDITION AT Dice.initWIthNWeapons"
	end

	# Description:
	#	Determines which player will start the match randomly.
	# Return
	#	Integer, from 0 to nPlayers-1
	def whoStarts(nPlayers)
		return @generator.rand(nPlayers)		
	end

	def firstShot
		if @generator.rand <= @FIRSTSHOTPROB
			return GameCharacter::SPACESTATION
		else
			return GameCharacter::ENEMYSTARSHIP
		end

		# Security check
		puts "WARNING! UNEXPECTED CONDITION AT Dice.firstShot"

	end
	
	# WIP
	def spaceStationMovers(speed)
		return @generator.rand < speed
	end
end

File: deepspace/src/Dice.java
/**
 * @author Sergio Quijano Rey
 * */

package Deepspace;

import java.util.Random;

/**
 * @brief Class to take all random decisions of the game
 * */
class Dice{
		// Private class attributes
		//======================================================================
		private final float NHANGARSPROB;
		private final float NSHIELDSPROB;
		private final float NWEAPONSPROB;
		private final float FIRSTSHOTPROB;

		// Private attributes
		//======================================================================
		private Random generator;

		// Constructors
		//======================================================================
		// WIP -- Deberia asignarse los valores constantes arriba, en la declaracion de los atributos
		Dice(){
				NHANGARSPROB = 0.25;
				NSHIELDSPROB = 0.25;
				NWEAPONSPROB = 0.33;
				FIRSTSHOTPROB = 0.5;

				generator = new Random(System.currentTimeMillis());
		}

		// Getters
		//======================================================================
		/**
		 * @brief determines the number of hangars that a space station receives at creation
		 * @return 	0, with NHANGARSPROB probability
		 * 			1, with 1 - NHANGARSPROB probability
		 * */
		public int initWithNHangars(){
				if(generator.nextFloat() < NHANGARSPROB){
						return 0;
				}else{
						return 1;
				}
		}

		/**
		 * @brief determines the number of weapons given to a space station on creation
		 * @return 	1, with NWEAPONSPROB probability
		 * 			2, with NWEAPONSPROB probability
		 * 			3, with 1 - 2*NWEAPONSPROB probability, that's to say, otherwise
		 * */
		public int initWithNWeapons(){
				float rand_value = generator.nextFloat();

				if(rand_value < NWEAPONSPROB){
						return 1;
				}else if(rand_value < 2*NWEAPONSPROB){
						return 2;
				}else{
						return 3;
				}

				System.out.println("WARNING! Unexpected condition at Dice.initWithNWeapons()");
				return 1;
		}

		/**
		 * @brief Determines the number of shield boosters given to a space station on creation
		 * @return 	0, with NSHIELDSPROB probability
		 * 			1, with 1 - NSHIELDSPROB probability
		 * */
		public int initWithNShields(){
				if(generator.nextFloat() < NSHIELDSPROB){
						return 0;
				}else{
						return 1;
				}

				System.out.println("WARNING! Unexpected condition at Dice.initWithNShields()");
				return 1;
		}

		/**
		 * @brief decides who starts the game
		 * @return an integer in the interval [0, nPlayers)
		 * */
		public int whoStarts(int nPlayers){
				return generator.nextInt(nPlayers);

		}

		/**
		 * @brief determines who shots first on a combat
		 * @return 	GameCharacter.SPACESTATION with a probability of FIRSTSHOTPROB
		 * 			GameCharacter.ENEMYSTARSHIP otherwise
		 * */
		public GameCharacter firstShot(){
				if(generator.nextFloat() < FIRSTSHOTPROB){
						return GameCharacter.SPACESTATION;
				}else{
						return GameCharacter.ENEMYSTARSHIP;
				}

				System.out.println("WARNING! Unexpected condition at Dice.firstShot()");
				return GameCharacter.SPACESTATION;
		}

		/**
		 * @brief determines if a spacestation moves to avoid a shoot
		 * @param speed: the current speed of the space station
		 * @return 	true, with a probability of speed
		 * 			false, otherwise
		 * */
		public boolean spaceStationMoves(float speed){
				if(generator.nextFloat() < speed){
						return true;
				}else{
						return false;
				}
		}
}
